packages/@aws-cdk/aws-s3/lib/bucket.ts

```diff
@@ -507,19 +514,51 @@ abstract class BucketBase extends Resource implements IBucket {
                 resourceArn: string, ...otherResourceArns: string[]) {
     const resources = [ resourceArn, ...otherResourceArns ];
 
-    const ret = iam.Grant.addToPrincipalOrResource({
-      grantee,
-      actions: bucketActions,
-      resourceArns: resources,
-      resource: this,
-    });
+    const crossAccountAccess = this.isGranteeFromAnotherAccount(grantee);
```

Comment:
I don't think it's possible to handle this in `addPrincipalAndResource`. The two branches differ in the `Grant` method called (`addToPrincipalAndResource` and `addToPrincipalOrResource`). Because of that, this can't be handled in a `Grant` method.

Reply:
Of course I missed this

OK, can you please add some comments explaining why each strategy is needed in each "if" branch?

Cognitive of these method names is too high for me. I literally stared at this for a good 40 seconds to try and figure out what's the difference between the code branches and they looked exactly the same.

Would have been so much more readable if the and/or distinction would have been a switch. Something like:

```ts
iam.Grant.add({
  grantee,
  actions: bucketActions,
  resourceArns: resources,
  resource: this,
  applyTo: iam.GrantTarget.PRINCIPAL_AND_RESOURCE // or iam.GrantTarget.PRINCIPAL_OR_RESOURCE
``

@rix0rrr is this something you can consider for #2344 ?

Reply:
Done.